### PR TITLE
Implemented Hive registration steps in Hive distcp.

### PR DESCRIPTION
**Pull request by ibuenros**

### CODE REVIEW

**Comment by ibuenros** [comment]
@ydai1124 can you review?

**Comment by ibuenros** [comment]
@zliu41 can you review?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveDataset.java [hunk 1]
```diff
@@ -119,244 +83,52 @@ public HiveDataset(FileSystem fs, HiveMetastoreClientPool clientPool, Table tabl
...
-      return new Path(pattern, this.table.getTableName());
-    }
+    log.info("Created Hive dataset for table " + tableIdentifier);
   }
 
   /**
    * Parse a location string from a {@link StorageDescriptor} into an actual list of globs.
    */
-  private List<Path> parseLocationIntoPaths(String sdLocation) {
+  protected static List<Path> parseLocationIntoPaths(Path sdLocation) {
```
Comment (ydai1124) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveDataset.java]: Is it possible to return multiple `Path`s?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveDataset.java [hunk 2]
```diff
@@ -119,244 +83,52 @@ public HiveDataset(FileSystem fs, HiveMetastoreClientPool clientPool, Table tabl
...
-  }
-
-  @Builder
-  private static class DiffPathSet {
-    @Singular(value = "copyFile") Collection<FileStatus> filesToCopy;
-    @Singular(value = "deleteFile") Collection<Path> pathsToDelete;
   }
 
-  private Map<List<String>, Partition> getPartitions(AutoReturnableObject<IMetaStoreClient> client, Table table)
+  protected static Map<List<String>, Partition> getPartitions(AutoReturnableObject<IMetaStoreClient> client, Table table)
```
Comment (ydai1124) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveDataset.java]: Why not moving these `static` methods to `utils`? They are generic and could be useful for others.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/entities/CommitStepCopyEntity.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * A {@link CopyEntity} encapsulating a {@link CommitStep}.
+ */
+public class CommitStepCopyEntity extends CopyEntity {
+
+  @Getter
+  private final CommitStep step;
+  @Getter
+  private final int priority;
```
Comment (ydai1124) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/entities/CommitStepCopyEntity.java]: Can you add javadoc for `priority`?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java [hunk 4]
```diff
@@ -0,0 +1,554 @@
+package gobblin.data.management.copy.hive;
```
Comment (ydai1124) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Missing license header.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveDataset.java [hunk 5]
```diff
@@ -70,47 +63,18 @@
...
-  private static final String tableToken = "$TABLE";
-
-  private final Properties properties;
-  private final FileSystem fs;
-  private final HiveMetastoreClientPool clientPool;
-  private final Table table;
-  private final Splitter splitter = Splitter.on(",").omitEmptyStrings();
+  protected final Properties properties;
+  protected final FileSystem fs;
+  protected final HiveMetastoreClientPool clientPool;
```
Comment (ydai1124) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveDataset.java]: Seems like it only used by `HiveCopyEntityHelper`. Do you still need this field here?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+
+    // All leftover paths in target are not in source, should be deleted.
+    for (Path deletePath : targetPathSet) {
+      builder.deleteFile(deletePath);
+    }
+
+    return builder.build();
+  }
+
+  private void checkTableCompatibility(Table sourceTable, Table targetTable) throws IOException {
```
Comment (ydai1124) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Why not return `boolean` instead of throwing `IOException` when checking compatibility?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java [hunk 7]
```diff
@@ -0,0 +1,554 @@
...
+
+      if (targetClient.get().tableExists(targetDatabase, this.dataset.table.getTableName())) {
+        this.existingTargetTable =
+            Optional.of(new Table(targetClient.get().getTable(targetDatabase, this.dataset.table.getTableName())));
+      } else {
+        this.existingTargetTable = Optional.absent();
+      }
+
+      if (this.existingTargetTable.isPresent()) {
+        checkTableCompatibility(this.dataset.table, this.existingTargetTable.get());
```
Comment (ydai1124) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Should `ExistingEntityPolicy` be used here to handle incompatible cases?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+        this.sourcePartitions = HiveDataset.getPartitions(sourceClient, this.dataset.table);
+        this.targetPartitions = this.existingTargetTable.isPresent() ?
+            HiveDataset.getPartitions(targetClient, this.existingTargetTable.get()) : Maps.<List<String>, Partition>newHashMap();
+      } else {
+        this.sourcePartitions = Maps.newHashMap();
+        this.targetPartitions = Maps.newHashMap();
+      }
+
+    } catch (TException te) {
+      throw new IOException("Failed to generate work units.", te);
```
Comment (ydai1124) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Can you modify the error message here?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java [hunk 9]
```diff
@@ -0,0 +1,554 @@
...
+    }
+  }
+
+  /**
+   * Finds all files read by the table and generates CopyableFiles. The semantics are as follows:
```
Comment (ydai1124) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Can you update the description here? It is not just `CopyableFiles`, but also other `CopyEntity`.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java [hunk 10]
```diff
@@ -0,0 +1,554 @@
...
+
+      if (targetPartition.isPresent()) {
+        targetPartitions.remove(partition.getValues());
+        try {
+          checkPartitionCompatibility(partition, targetPartition.get());
+        } catch (IOException ioe) {
+          if (existingEntityPolicy == ExistingEntityPolicy.ABORT) {
+            log.error("Source and target partitions are not compatible. Aborting copy of partition " + this.partition);
+            return Lists.newArrayList();
+          }
```
Comment (ydai1124) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Do you need to check if `ExistingEntityPolicy` equals to `REPLACE`? It is dangerous to make `REPLACE` default here.

File: gobblin-utility/src/main/java/gobblin/util/io/GsonInterfaceAdapter.java [hunk 11]
```diff
@@ -48,45 +60,119 @@
...
+      // delegate primitives, arrays, collections, and maps
+      return null;
     }
+    if (!this.baseClass.isAssignableFrom(type.getRawType())) {
+      // delegate anything not assignable from base class
+      return null;
+    }
+    TypeAdapter<R> adapter = new InterfaceAdapter<>(gson, this, type);
+    return adapter;
   }
```
Comment (ydai1124) [inline gobblin-utility/src/main/java/gobblin/util/io/GsonInterfaceAdapter.java]: Can you add some description about unable to serialize `generic`?

File: gobblin-utility/src/test/java/gobblin/util/io/GsonInterfaceAdapterTest.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
+package gobblin.util.io;
```
Comment (ydai1124) [inline gobblin-utility/src/test/java/gobblin/util/io/GsonInterfaceAdapterTest.java]: Can you add `@Test groups` and license header to the newly added tests?

File: gobblin-hive-registration/src/main/java/gobblin/hive/metastore/HiveMetaStoreUtils.java [hunk 13]
```diff
@@ -101,6 +101,15 @@ public static HiveTable getHiveTable(Table table) {
   }
 
   /**
+   +   * Covert a {@link Table} into a {@link HiveTable} with customized value for {@link HiveConstants#LOCATION}.
+   +   */
+  public static HiveTable getHiveTableWithNewLocation(Table table, String location) {
```
Comment (ydai1124) [inline gobblin-hive-registration/src/main/java/gobblin/hive/metastore/HiveMetaStoreUtils.java]: Can you remove these two methods since they are not used any more?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/entities/CommitStepCopyEntity.java (hunk 3 above)
Comment (ydai1124) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/entities/CommitStepCopyEntity.java]: Should `priority` be immutable?

File: gobblin-data-management/src/main/java/gobblin/util/commit/DeleteFileCommitStep.java [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+import com.google.common.base.Function;
+import com.google.common.collect.Lists;
+
+import gobblin.commit.CommitStep;
+import gobblin.data.management.trash.Trash;
+import gobblin.data.management.trash.TrashFactory;
+
+
+/**
+ * {@link CommitStep} to delete a set of paths in HDFS.
```
Comment (ydai1124) [inline gobblin-data-management/src/main/java/gobblin/util/commit/DeleteFileCommitStep.java]: nit: HDFS => FileSystem

File: gobblin-hive-registration/src/main/java/gobblin/hive/PartitionDeregisterStep.java [hunk 15]
```diff
@@ -0,0 +1,38 @@
+package gobblin.hive;
```
Comment (ydai1124) [inline gobblin-hive-registration/src/main/java/gobblin/hive/PartitionDeregisterStep.java]: Missing license header.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveDataset.java (hunk 1 above)
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveDataset.java]: removed this method. It turns out the location is a `Path`, but `Path` is a very generic thing in Hadoop, and it can include globs, etc.
Instead of doing this, I am just checking that path is not a glob.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/entities/CommitStepCopyEntity.java (hunk 3 above)
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/entities/CommitStepCopyEntity.java]: I think it should be immutable. Do you disagree?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java (hunk 4 above)
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Fixed

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveDataset.java (hunk 5 above)
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveDataset.java]: Needs to be here because it is supplied by the dataset finder, and the `HiveDataset` passes it to the helper at a later time.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java (hunk 6 above)
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Because there is more information in an `IOException`, including a message with the precise reason and a stack trace.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java (hunk 7 above)
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Don't want to do that right now, it seems that replacing a table is a riskier operation than replacing partitions.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java (hunk 9 above)
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Fixed

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java (hunk 10 above)
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java]: Fixed

File: gobblin-utility/src/main/java/gobblin/util/io/GsonInterfaceAdapter.java (hunk 11 above)
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/io/GsonInterfaceAdapter.java]: Fixed

File: gobblin-hive-registration/src/main/java/gobblin/hive/metastore/HiveMetaStoreUtils.java (hunk 13 above)
Comment (ibuenros) [inline gobblin-hive-registration/src/main/java/gobblin/hive/metastore/HiveMetaStoreUtils.java]: Fixed

File: gobblin-data-management/src/main/java/gobblin/util/commit/DeleteFileCommitStep.java (hunk 14 above)
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/util/commit/DeleteFileCommitStep.java]: Fixed

File: gobblin-hive-registration/src/main/java/gobblin/hive/PartitionDeregisterStep.java (hunk 15 above)
Comment (ibuenros) [inline gobblin-hive-registration/src/main/java/gobblin/hive/PartitionDeregisterStep.java]: Fixed

File: gobblin-utility/src/test/java/gobblin/util/io/GsonInterfaceAdapterTest.java (hunk 12 above)
Comment (ibuenros) [inline gobblin-utility/src/test/java/gobblin/util/io/GsonInterfaceAdapterTest.java]: I added the license headers. Don't quite understand what you mean by `@Test groups`.

**Comment by ibuenros** [comment]
@ydai1124 fixed comments, and did some changes to code.

File: gobblin-hive-registration/src/main/java/gobblin/hive/metastore/HiveMetaStoreBasedRegister.java [hunk 16]
```diff
@@ -127,52 +133,41 @@ private boolean createDbIfNotExists(IMetaStoreClient client, String dbName) thro
...
-    IMetaStoreClient client = borrowClient();
 
-    HiveLock lock = locks.getTableLock(table.getDbName(), table.getTableName());
-    lock.lock();
-    try {
+    try(AutoReturnableObject<IMetaStoreClient> client = this.clientPool.getClient();
+        HiveLock lock = locks.getTableLock(table.getDbName(), table.getTableName())) {
       try {
-        client.getPartition(table.getTableName(), table.getDbName(), partition.getValues());
+        client.get().getPartition(table.getTableName(), table.getDbName(), partition.getValues());
```
Comment (zliu41) [inline gobblin-hive-registration/src/main/java/gobblin/hive/metastore/HiveMetaStoreBasedRegister.java]: 1st parameter is dbname

File: gobblin-config-management/gobblin-config-core/src/main/java/gobblin/config/store/hdfs/SimpleHDFSConfigStore.java [hunk 17]
```diff
@@ -275,7 +275,7 @@ public URI getStoreURI() {
       if (!includesFileStatus.isDir()) {
         try (InputStream includesConfInStream = this.fs.open(includesFileStatus.getPath())) {
           configKeyPaths.addAll(Lists.newArrayList(
-              Iterables.transform(IOUtils.readLines(includesConfInStream, Charsets.UTF_8), new IncludesToConfigKey())));
+              Iterables.transform(IOUtils.readLines(includesConfInStream, Charsets.UTF_8.name()), new IncludesToConfigKey())));
```
Comment (zliu41) [inline gobblin-config-management/gobblin-config-core/src/main/java/gobblin/config/store/hdfs/SimpleHDFSConfigStore.java]: is this change necessary?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveUtils.java [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+  public static Set<Path> getPaths(InputFormat<?, ?> inputFormat, Path location) throws IOException {
+    JobConf jobConf = new JobConf(getHadoopConfiguration());
+
+    Set<Path> paths = Sets.newHashSet();
+
+    FileInputFormat.addInputPaths(jobConf, location.toString());
+    InputSplit[] splits = inputFormat.getSplits(jobConf, 1000);
+    for (InputSplit split : splits) {
+      if (!(split instanceof FileSplit)) {
+        throw new IOException("Not a file split.");
```
Comment (zliu41) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveUtils.java]: can you include the class of the split in the exception msg

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java [hunk 19]
```diff
@@ -181,6 +197,37 @@ private void publishFileSet(CopyEntity.DatasetAndPartition datasetAndPartition,
...
+    for (WorkUnitState wus : workUnits) {
+      if (baseClass.isAssignableFrom(CopySource.copyEntityClass(wus))) {
+        CommitStepCopyEntity step = (CommitStepCopyEntity) CopySource.deserializeCopyEntity(wus);
+        steps.add(step);
+      }
+    }
+
+    Comparator<CommitStepCopyEntity> commitStepSorter = new Comparator<CommitStepCopyEntity>() {
+      @Override public int compare(CommitStepCopyEntity o1, CommitStepCopyEntity o2) {
+        return Integer.compare(o1.getPriority(), o2.getPriority());
```
Comment (zliu41) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java]: this is sorting from low priority to high priority?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriter.java [hunk 20]
```diff
@@ -177,8 +177,9 @@ protected static Path getPartitionOutputRoot(Path outputDir,
 
   public static Path getOutputFilePath(CopyableFile file, Path outputDir,
       CopyEntity.DatasetAndPartition datasetAndPartition) {
+    Path destinationWithoutSchemeAndAuthority = PathUtils.getPathWithoutSchemeAndAuthority(file.getDestination());
```
Comment (zliu41) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriter.java]: Why not just use `path.toUri().getRawPath()`

File: gobblin-data-management/src/main/java/gobblin/util/commit/DeleteFileCommitStep.java [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+
+  private static List<FileStatus> toFileStatus(FileSystem fs, Collection<Path> paths) throws IOException {
+    List<FileStatus> fileStatuses = Lists.newArrayList();
+    for (Path path : paths) {
+      fileStatuses.add(fs.getFileStatus(path));
+    }
+    return fileStatuses;
+  }
+
+  @Override public boolean isCompleted() throws IOException {
```
Comment (zliu41) [inline gobblin-data-management/src/main/java/gobblin/util/commit/DeleteFileCommitStep.java]: Is this method used?

File: gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegisterStep.java [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+  private final HiveRegProps props;
+
+  @Override public boolean isCompleted() throws IOException {
+    // TODO: this is complicated due to preactivities, postactivities, etc. but unnecessary for now because exactly once
+    // is not enabled.
+    return false;
+  }
+
+  @Override public void execute() throws IOException {
+    HiveRegister hiveRegister = new HiveMetaStoreBasedRegister(this.props, this.metastoreURI);
```
Comment (zliu41) [inline gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegisterStep.java]: can you avoid binding to `HiveMetaStoreBasedRegister` here, and use `HiveRegister.get`

File: gobblin-hive-registration/src/main/java/gobblin/hive/HiveStripedLocks.java [hunk 23]
```diff
@@ -53,24 +53,28 @@ public HiveLock getTableLock(String dbName, String tableName) {
    *
    * Supports operations {@link #lock()} and {@link #unlock()}.
    */
-  public static class HiveLock {
+  public static class HiveLock implements AutoCloseable {
```
Comment (zliu41) [inline gobblin-hive-registration/src/main/java/gobblin/hive/HiveStripedLocks.java]: can you update javadoc

File: gobblin-hive-registration/src/main/java/gobblin/hive/metastore/HiveMetaStoreBasedRegister.java [hunk 24]
```diff
@@ -64,47 +66,45 @@
...
   private final HiveStripedLocks locks = new HiveStripedLocks();
 
   public HiveMetaStoreBasedRegister(State state) throws IOException {
+    this(state, Optional.<String>absent());
+  }
+
+  public HiveMetaStoreBasedRegister(State state, Optional<String> metastoreURI) throws IOException {
     super(state);
 
     GenericObjectPoolConfig config = new GenericObjectPoolConfig();
```
Comment (zliu41) [inline gobblin-hive-registration/src/main/java/gobblin/hive/metastore/HiveMetaStoreBasedRegister.java]: is this still used?

File: gobblin-hive-registration/src/main/java/gobblin/hive/spec/SimpleHiveSpec.java [hunk 25]
```diff
@@ -30,26 +39,49 @@
...
   }
 
-  @Getter
   public static class Builder<T extends Builder<?>> {
-    protected final Path path;
-    protected HiveTable table;
-    protected Optional<HivePartition> partition = Optional.absent();
+    private Path path;
+
+    private Builder() {
```
Comment (zliu41) [inline gobblin-hive-registration/src/main/java/gobblin/hive/spec/SimpleHiveSpec.java]: why do you need this constructor?

File: gobblin-utility/src/main/java/gobblin/util/io/GsonInterfaceAdapter.java [hunk 26]
```diff
@@ -12,81 +12,181 @@
...
-
-  @Override
-  public JsonElement serialize(T object, Type interfaceType, JsonSerializationContext context) {
-    JsonObject wrapper = new JsonObject();
-    wrapper.addProperty(OBJECT_TYPE, object.getClass().getName());
-    wrapper.add(OBJECT_DATA, PURE_GSON.toJsonTree(object));
-    return wrapper;
-  }
+@RequiredArgsConstructor
+public class GsonInterfaceAdapter implements TypeAdapterFactory {
```
Comment (zliu41) [inline gobblin-utility/src/main/java/gobblin/util/io/GsonInterfaceAdapter.java]: Can you summarize what's the purpose of the changes to this class.

File: gobblin-hive-registration/src/main/java/gobblin/hive/metastore/HiveMetaStoreBasedRegister.java (hunk 16 above)
Comment (ibuenros) [inline gobblin-hive-registration/src/main/java/gobblin/hive/metastore/HiveMetaStoreBasedRegister.java]: It is not actually, if the last parameter is a `String`, the order is dbname, tablename, if the last parameter is a `List<String>`, the order is tablename, dbname. (They really should have thought better about the API).

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java (hunk 19 above)
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java]: This is the same order that a priority queue would generate. I clarified in the javadoc in `CommitStepCopyEntity`.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriter.java (hunk 20 above)
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriter.java]: `PathUtils.withoutLeadingSeparator` takes a `Path` as input, so I need a `Path` either way.

File: gobblin-data-management/src/main/java/gobblin/util/commit/DeleteFileCommitStep.java (hunk 21 above)
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/util/commit/DeleteFileCommitStep.java]: Not for now, but it is part of the `CommitStep` API

File: gobblin-hive-registration/src/main/java/gobblin/hive/spec/SimpleHiveSpec.java (hunk 25 above)
Comment (ibuenros) [inline gobblin-hive-registration/src/main/java/gobblin/hive/spec/SimpleHiveSpec.java]: Because of lombok. Otherwise lombok would create a public, no arguments, constructor.

File: gobblin-utility/src/main/java/gobblin/util/io/GsonInterfaceAdapter.java (hunk 26 above)
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/io/GsonInterfaceAdapter.java]: In the previous version calling `GsonInterfaceAdapter.getGson(MyClass.class).toJson(new MyClass)` would create an infinite loop, so this method was useful mostly for interfaces, and not concrete classes. Also `GsonInterfaceAdapter.getGson(MyClass.class).toJson(new MyClassDerived)` would not actually write the class name, which made it very hard to correctly serialize classes with internal polymorphic fields.
The changes also add support for `Optional`.

**Comment by ibuenros** [comment]
@zliu41 Addressed comments.

File: gobblin-utility/src/test/java/gobblin/util/io/GsonInterfaceAdapterTest.java (hunk 12 above)
Comment (ydai1124) [inline gobblin-utility/src/test/java/gobblin/util/io/GsonInterfaceAdapterTest.java]: I mean the testng groups, e.g., @Test(groups = {"gobblin.util.io"})

**Comment by ydai1124** [comment]
LGTM other than one more comment.

File: gobblin-hive-registration/src/main/java/gobblin/hive/metastore/HiveMetaStoreBasedRegister.java (hunk 16 above)
Comment (zliu41) [inline gobblin-hive-registration/src/main/java/gobblin/hive/metastore/HiveMetaStoreBasedRegister.java]: No, `getPartition(String db_name, String tbl_name, List<String> part_vals)` in `HiveMetaStoreClient` overrides `getPartition(String tblName, String dbName, List<String> partVals)` in `IMetaStoreClient`.

It's ok if you leave it this way since I'll make the change anyway in my pr

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java (hunk 19 above)
Comment (zliu41) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java]: But are you executing lower priority steps first? Shouldn't you do the reverse?

Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java]: Steps with lower priority value will execute first, yes. This is the same way that a priority queue would return them (correct me if I'm wrong).

File: gobblin-hive-registration/src/main/java/gobblin/hive/metastore/HiveMetaStoreBasedRegister.java (hunk 16 above)
Comment (ibuenros) [inline gobblin-hive-registration/src/main/java/gobblin/hive/metastore/HiveMetaStoreBasedRegister.java]: I just saw that, you're right. Hive API is pretty dumb.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java [hunk 27]
```diff
@@ -147,11 +156,18 @@ private void publishFileSet(CopyEntity.DatasetAndPartition datasetAndPartition,
...
-        metadata.getDatasetURN()));
+    log.info(String.format("[%s] Publishing fileSet from %s for dataset %s", datasetAndPartition.identifier(),
+        datasetWriterOutputPath, metadata.getDatasetURN()));
 
+    List<CommitStep> prePublish = getCommitSequence(datasetWorkUnitStates, PrePublishStep.class);
+    List<CommitStep> postPublish = getCommitSequence(datasetWorkUnitStates, PostPublishStep.class);
+    log.info(String.format("[%s] Found %d prePublish steps and %d postPublish steps.", datasetAndPartition.identifier(),
+        prePublish.size(), postPublish.size()));
+
+    executeCommitSequence(prePublish);
```
Comment (zliu41) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java]: What is the consequence of failing to execute these steps? Is it going to cause loss of data? Duplicate data? Or no problem?

Since commit steps and commit sequences were created for exactly-once semantics (at least the javadoc says so), if you are using them for different purposes can you either add some comments or modify the javadoc of commit step and commit seuqence.

Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java]: It can cause any of that, same as any failed publish step.
I am using `CommitStep` here with the objective of eventually using exactly-once semantics (very easy to assemble a `CommitSequence` if we already have the steps). However, I understand exactly-once is not ready yet.

**Comment by zliu41** [comment]
LGTM
